Sources/GRPCCore/Transport/InProcessClientTransport.swift

```diff
@@ -0,0 +1,282 @@
+          state = .connected(serverTransport, openStreams + 1, signalEndConnection)
+        } catch let acceptStreamError as RPCError {
+          return .failure(acceptStreamError)
+        } catch {
+          return .failure(RPCError(code: .unknown, message: "Unknown error: \(error)."))
+        }
+        return .success(.success)
+
+      case .unconnected(let serverTransport, var pendingStreams):
+        let (stream, continuation) = AsyncStream<Void>.makeStream()
+        pendingStreams.append(continuation)
+        state = .unconnected(serverTransport, pendingStreams)
+        return .success(.pending(stream))
+
+      case .closed:
+        return .failure(
+          RPCError(
+            code: .failedPrecondition,
+            message: "The client transport is closed."
+          )
+        )
+      }
+    }
+
+    let withStreamResult: WithStreamResult
+    do {
+      withStreamResult = try result.get()
+    } catch {
+      serverStream.outbound.finish()
+      clientStream.outbound.finish()
+      throw error
+    }
+
+    switch withStreamResult {
+    case .success:
+      ()
+    case .pending(let pendingStream):
+      for await _ in pendingStream {
+        // This loop will exit either when the task is cancelled or when the
+        // client connects and this stream can be opened.
+      }
+      try Task.checkCancellation()
+
+      try self.state.withLockedValue { state in
+        switch state {
+        case .unconnected:
+          fatalError("Invalid state.")
+        case .connected(let serverTransport, let openStreams, let signalEndConnection):
+          do {
+            try serverTransport.acceptStream(serverStream)
+            state = .connected(serverTransport, openStreams + 1, signalEndConnection)
+          } catch let acceptStreamError as RPCError {
+            throw acceptStreamError
+          } catch {
+            throw RPCError(code: .unknown, message: "Unknown error: \(error).")
+          }
+        case .closed:
+          serverStream.outbound.finish()
+          clientStream.outbound.finish()
+          throw RPCError(
+            code: .failedPrecondition,
+            message: "The client transport is closed."
+          )
+        }
+      }
+    }
+
+    defer {
+      self.state.withLockedValue { state in
+        switch state {
+        case .unconnected:
+          fatalError("Invalid state")
+        case .connected(let serverTransport, let openStreams, let signalEndContinuation):
+          state = .connected(serverTransport, openStreams - 1, signalEndContinuation)
+        case .closed(let openStreams, let signalEndConnection):
+          if openStreams == 1 {
+            // This was the last open stream: signal the closure of the client.
+            signalEndConnection?.finish()
+          }
+        }
+      }
+    }
+
+    let userResult = try await closure(clientStream)
+
+    serverStream.outbound.finish()
+    clientStream.outbound.finish()
```

Comment:
Should we be doing this in the `defer` instead?

Reply:
Yeah you're right, good catch